src/tools/grpc_server.cc

```diff
@@ -235,8 +290,58 @@ Status CodeSearchImpl::Search(ServerContext* context, const ::Query* request, ::
         return Status(StatusCode::INVALID_ARGUMENT, "Parse error");
     }
 
+    string line_pat = q.line_pat->pattern();
+
+    /* Patterns like "User.*Info" and "p\d+" might be genuine attempts
+       to match tags, so we cannot too quickly dismiss odd-looking REs
+       as justifying our skipping the phases of a tags search.  But we
+       can at least special-case a few characters that would not ever
+       appear in a pattern that matches tags.
+       TODO(brandon-rhodes): make this more sophisticated. */
+    bool might_match_tags =
+        // Characters that can't appear in an RE that matches a tag.
+        line_pat.find_first_of(" !\"#%&',-/;<=>@`") == string::npos
+        // If the user anchored the RE, then we can only run it against
+        // whole lines from the corpus, never against tags.
+        && line_pat.front() != '^'
+        && line_pat.back() != '$'
+        ;
+
     match_stats stats;
-    if (q.tags_pat == NULL) {
+    if (q.tags_pat == NULL && tagdata_ && might_match_tags) {
+        string regex;
+        int32_t max_matches = q.max_matches;  // remember original value
+
+        add_match::line_set unique_lines;
+        add_match cb(&unique_lines, response);
+
+        /* To surface the most important matches first, start with tags.
+           First pass: is the pattern an exact match for any tags? */
+        regex = "^" + line_pat + "$";
+        q.line_pat.reset(new RE2(regex, q.line_pat->options()));
+        run_tags_search(q, tagdata_, cb, tagmatch_, stats);
+
+        q.max_matches = max_matches - cb.match_count();
+        if (q.max_matches > 0) {
+
+            /* Second pass: is the pattern a prefix match for any tags? */
+            regex = "^" + line_pat + "[^\t]";
+            q.line_pat.reset(new RE2(regex, q.line_pat->options()));
+            run_tags_search(q, tagdata_, cb, tagmatch_, stats);
+
+            q.max_matches = max_matches - cb.match_count();
+            if (q.max_matches > 0) {
+
+              /* Third and final pass: full corpus search. */
+              q.line_pat.reset(new RE2(line_pat, q.line_pat->options()));
+              code_searcher::search_thread *search;
+              if (!pool_.try_pop(&search))
+                search = new code_searcher::search_thread(cs_);
+              search->match(q, cb, cb, &stats);
```

Comment:
Hm, this is maybe minor, but we're going to lose accuracy on `stats` here; the last search will win. I think the thing I might suggest is:

- add a default constructor for `stats` that zeros out all the fields
- delete the [`memset`](https://github.com/livegrep/livegrep/blob/0eeba4bcbd7d73f75288dc46bd9cfebb1c887196/src/codesearch.cc#L1150) here
- make [this code](https://github.com/livegrep/livegrep/blob/0eeba4bcbd7d73f75288dc46bd9cfebb1c887196/src/codesearch.cc#L1166-L1174) and [this code](https://github.com/livegrep/livegrep/blob/0eeba4bcbd7d73f75288dc46bd9cfebb1c887196/src/codesearch.cc#L246-L252) *add* to `stats`, instead of setting.

that will let us accumulate counters with no additional work

Reply:
:+1: